sky/spot/controller.py

```diff
@@ -6,6 +6,7 @@
 
 import colorama
 import filelock
+import ray
```

Comment:
Out of curiosity - any particular reason to use Ray? Could this have been implemented with `multiprocessing`?

Reply:
Ahh, there is no particular reason for implementing it with `ray`. The only reason is that I am more familiar with the busy loop with `ray.wait`, and our VMs already have a ray cluster running in the background.

Do you think it would be better to implement it with `multiprocessing`?

Reply:
I think multiprocessing might be faster, but ray is also fine if its faster for us to implement.

Reply:
Just tried out `multiprocessing`, `concurrent.futures` and `asyncio`, they both have the functionality to support running the process in parallel, but they all have drawbacks:
1. `multiprocessing.Process`: We can start the process and check the liveness of the process with `is_alive`. We can cancel the process with `os.kill(p.pid, signal.SIGINT)`. However, we cannot easily catch the exceptions happening inside the process from the main process. Additional codes are needed to do so, as mentioned [here](https://stackoverflow.com/a/63773140).
2. `concurrent.futures`: We can start the process with the `Executor.submit` and catch the exceptions from the main process, but the future object does not have a way to stop the running process, unlike `ray.cancel`.
3. `asyncio.create_task`: Based on their [doc](https://docs.python.org/3/library/asyncio-task.html#asyncio.Task.cancel), the task is not guaranteed to be cancelled, but we need that guarantee to cancel the `_controller_run` function when the `CANCEL` signal received.  

Based on the drawback above, I think it might be fine to use `ray` to handle the async and future as we did in the code. Wdyt @romilbhardwaj ?

Reply:
Ah for `multiprocessing`, I was thinking of using `multiprocessing.Pool` since it provides [nice exception handling](https://stackoverflow.com/a/8708806), but turns out its not easy to cancel [running processes](https://stackoverflow.com/a/35807790) in it. 

I think it's fine to use `ray` for now. Thanks for the extensive investigation into different methods to do this!